src/nccl_ofi_rdma.c

```diff
@@ -5265,47 +5293,61 @@ static int post_eager_copy(nccl_net_ofi_rdma_req_t *req)
 static int post_flush_req(nccl_net_ofi_rdma_req_t *req)
 {
  	nccl_net_ofi_rdma_recv_comm_t *r_comm = (nccl_net_ofi_rdma_recv_comm_t *)req->comm;
+	nccl_net_ofi_rdma_ep_t *ep = (nccl_net_ofi_rdma_ep_t *)r_comm->base.base.ep;
+	nccl_net_ofi_rdma_flush_buffer_t *f_buff = &r_comm->flush_buff;
 	rdma_req_flush_data_t *flush_data = get_flush_data(req);
-	nccl_net_ofi_schedule_t *schedule = flush_data->schedule;
+	nccl_net_ofi_schedule_t *schedule;
+	nccl_net_ofi_xfer_info_t *xfer_info;
+	nccl_net_ofi_rdma_recv_comm_rail_t *comm_rail;
+	ssize_t rc = 0;
 
-	assert(schedule != NULL);
+	assert(((f_buff->cpu_cache_line_size * ep->num_rails) + r_comm->flush_buff.size) <= f_buff->size);
```

Comment:
One more thing... I don't understand this assert.

```
f_buff == &r_comm->flush_buff
```
so
```
f_buff->size == r_comm->flush_buff.size
```
and this assertion is
```
assert(<positive thing> + r_comm->flush_buff.size <= f_buff->size);
```
Seems like it could never be true?

Reply:
@rauteric thanks good catch. `f_buff->size` has to be replaced by `system_page_size`. Need to check how it bails out assert() check in my first test though.

Reply:
Why does there need to be any assert? What possible condition is this attempting to catch?

Reply:
it needs to catch the case of unexpected cache line size and number of rails, which are not cross-checked in any other places.

Reply:
It's already been called out that this should probably occur once during init and not in the datapath, but going a step further, why rails, why cache line size, and why the product of the two? Is there a coherency concern?

Reply:
Really, page aligned might be the right answer, but would mean more complicated code to keep memory usage sane.  The code is trying to avoid write conflicts from the 4 devices, as they're all likely to complete at the same time.

Reply:
> The code is trying to avoid write conflicts from the 4 devices, as they're all likely to complete at the same time.

Ah, I missed this line:

```
f_buff->host_buffer + (f_buff->cpu_cache_line_size * rail_seq);
```

I didn't see any other usage besides this assert, apologies.

Still, this feels like a really backwards way of doing this. What you really want is just to add alignment attributes. https://godbolt.org/z/aa36s1Yhe

This ensures

1. All are on a common page.
2. each rail is reading an independent cache line on that page.
3. size is as minimized as is possible.

Same exact thing can be expressed as a flexible array in C, with `__attribute__((aligned(64)))`; Although you still need num_rails to be a compile-time constant to be able to compute the overall alignment requirement and ensure it's all on a common page.

> it needs to catch the case of unexpected cache line size and number of rails, which are not cross-checked in any other places.

It should not be possible AFAIK to allocate any struct on a freelist exceeding a page size today, because we maintain the metadata on the same page and we won't allow page splits. So I'd expect that an assert in the freelist allocation would always be hit prior to this check.

Reply:
Oh wait, this can't work, because our freelist implementation casts to `void*` and there's no valid mechanism to query the alignment of the type of the thing we're allocating. WCGW.

Reply:
on the other hand, there's no reason for this buffer to be on a free list.

Reply:
not something introduced by this pr, but why is:

```
struct nccl_net_ofi_rdma_recv_comm {
        [ ... ]
	/* The flush buffer */
	nccl_net_ofi_rdma_flush_buffer_t flush_buff;
        [ ... ]
};
```

safe in the first place? Only looked briefly and not an issue introduced in this PR, but I see absolutely no check for the invalidity of the comm's flush_buff before we overwrite it with a new one.

Where does nccl guarantee that it won't post data ops, post a flush to sync them, post an independent set of data ops that have no coherency requirements relative to the previous set, post a second flush for the second set, and then tests the two in parallel?